src/log_batch.rs

```diff
@@ -550,50 +542,90 @@ where
             return None;
         }
 
-        // layout = { 8 bytes len | item count | multiple items | 4 bytes checksum }
-        let mut vec = Vec::with_capacity(4096);
-        let mut entries_size = 0;
-        vec.encode_u64(0).unwrap();
-        vec.encode_var_u64(self.items.len() as u64).unwrap();
+        // layout = { 8 bytes batch_len | 8 bytes section offset | item count | multiple items (without entries) | crc32 | section (entries) | crc32 }
```

Comment:
```suggestion
        // Log Batch Layout:
        // <---    header     ---> <---       entries        ---> <---                   footer                    --->
        // { u32 len | u32 offset | (compressed) entries | crc32 | item count | entry indexes / kvs / commands | crc32 }
        //                    `------------------------------------^
```

Reply:
To avoid using two buffers, you can encode entries first (and update `EntryIndex` struct), then encode entry indexes and other items in the second pass.

Reply:
BTW, during recovery, it's more economical to read the footer of current batch and the header of next batch together.

Reply:
> To avoid using two buffers, you can encode entries first (and update EntryIndex struct), then encode entry indexes and other items in the second pass.

Yep, that's my first design. I didn't implement like that becase when recovering a log batch, we need to read the header part and the entries_index part. That design will make utilizing cache a little bit completed before (before I implemented `LogBatchFileReader`).

Reply:
Not sure I understand, but "read the footer of current batch and the header of next batch together" can is as cache friendly as current implementation.